exercises/binary-search/example.cpp

```diff
@@ -0,0 +1,28 @@
+#include "binary_search.h"
+
+namespace binary_search {
+
+int find(std::vector<int> const& data, int value)
+{
+    int lpos = 0;
+    int rpos = data.size() - 1;
+    int mid = 0;
+
+    while (lpos <= rpos)
+    {
+        mid = (lpos + rpos)/2;
```

Comment:
There is a bug on this line, by the way. Though once you change the variables to `size_t` it will effectively never be hit.

Reply:
Sorry I should've been clear, I said this in a 'this is an interesting thing I've come across' way and not in a 'go fix this weird bug' way.

I like binary search, and I think it's a neat fact that it was first proposed in the 1940's but the first correct implementation didn't appear until the 1960's. And [most implementations are wrong due to the exact bug here](https://ai.googleblog.com/2006/06/extra-extra-read-all-about-it-nearly.html).

Also, I work in the optimization space, and I've used some of the ideas in [this paper](https://arxiv.org/abs/1509.05053) in interviews to explain some of the kinds of analysis work that I do. Particularly interesting are the interactions between the branchy and branch-free binary search algorithms with the cache.

Reply:
Thanks for the explanations, the reading suggestions and for pointing out the bug in the first place! The first short article was a really enlightening read. Will defo put some time aside for a cup of tea and the arXiv paper tomorrow.

More of this, please! :-)